### ISSUE TITLE
Optimize test parallelization

### ISSUE THREAD

**Issue by eloquence** [freedomofpress/securedrop#6026]
`app-tests` currently uses a parallelism factor of 3:
https://github.com/freedomofpress/securedrop/blob/14c90e0e32d5c99af0423671fc87e655ac33c1fd/.circleci/config.yml#L91

We are observing job runs with run times like 20 minutes/40 minutes/20 minutes. We should consider if increasing the parallelism or optimizing test distribution could reduce overall job run time further.

**Comment by cfm** [comment]
On preliminary investigation:  For a job `n`, CircleCI is *storing* pytest timings in each run's `junit.xml` artifact, but it's not able to *follow* these timings on job `n+1` ([e.g.](https://app.circleci.com/pipelines/github/freedomofpress/securedrop/2793/workflows/871935ec-36ac-4d2d-a7ac-d63f51a1de15/jobs/55896/steps?invite=true#step-109-17)):

```
Error autodetecting timing type, falling back to weighting by name. Autodetect no matching filename or classname.  If file names are used, double check paths for absolute vs relative.
Example input file: "tests/functional/test_admin_interface.py"
Example file from timings: ""
```

It looks like `pytest` is identifying tests in `junit.xml` only by the `classname` attribute, whereas CircleCI expects to find them identified by `file`, which is why it's (effectively) not actually learning tests' timings and rebalancing their parallelism.  See also:

* https://support.circleci.com/hc/en-us/articles/360056636191-Common-test-splitting-issues-with-pytest
* https://discuss.circleci.com/t/circleci-tests-split-does-not-work-as-advertised/38111

I'll discuss picking this up with @zenmonkeykstop tomorrow.

**Comment by cfm** [comment]
While we wait to see what timings `app-tests` stabilizes on after #6065, I'm experimenting with other test-splitting strategies in `6026-parallelism-*` branches.  Each will need to be run twice: a first time to generate timings and a second time to consume them and see what changes.

**Comment by cfm** [comment]
**Tl;dr:**  The current effective lower bound on `app-tests`'s runtime is the runtime of `tests/pageslayout/test_journalist.py`.  We can optimize down to that lower bound by either:

1. factoring out page-layout tests into their own job explicitly with `parallelism: 2`; or
2. running `app-tests` with a `parallelism` sufficient to put `tests/pageslayout/test_journalist.py` implicitly into its own run.

I'll open a pull request to propose (1) for the sake of actionable discussion, but I'm happy to field questions, feedback, and higher-level discussion here too.

Further optimization below this lower bound will require either:

1. optimizing the performance of the page-layout tests themselves; or
2. refactoring tests into smaller classes and/or files on which CircleCI can split tests more finely across runs.


---

I've tested the following scenarios:

| Scenario | Branch | Best Runtime | Notes |
| --- | --- | --- | --- |
| *Control* | `develop` | [58:14](https://app.circleci.com/pipelines/github/freedomofpress/securedrop/2876/workflows/772748fe-3f35-48c4-92a5-1514ab28822c) | This was the most-recent `app-tests` job when I began this experiment on Monday. |
| 5x parallelism | `6026-parallelism-5` | [33:07](https://app.circleci.com/pipelines/github/freedomofpress/securedrop/2877/workflows/5b7ad5c4-b993-41a4-95e0-6948dd144a74) | Failed consistently on `test_worker_for_job`. |
| 10x parallelism| `6026-parallelism-10` | [40:11](https://app.circleci.com/pipelines/github/freedomofpress/securedrop/2883/workflows/57673efa-593c-4f92-876c-d9ff60d2435c) | [Another run](https://app.circleci.com/pipelines/github/freedomofpress/securedrop/2883/workflows/abe7c3ad-23ae-4955-acd1-2c6e7a7c2873) claims 30:45, but this can't be the case for `lint` (8:25) + `app-tests` (30:42).... |
| Factor out page-layout tests into separate job | `6026-parallelism-split-pageslayout` | [39:47](https://app.circleci.com/pipelines/github/freedomofpress/securedrop/2885/workflows/4e6fbad8-ef90-4c4d-993a-b9689b988cc1) | HTML validation was broken; fixed in a [subsequent run](https://app.circleci.com/pipelines/github/freedomofpress/securedrop/2890/workflows/c3bfa684-d5a3-4e1d-95ed-47917651af9c) with a longer `lint` job. |

Together, these runs reveal an implication I hadn't realized in the course of #6065.  Although pytest reports timings per test (function), CircleCI is able to split tests for parallelism only per filename[1], *not* per function.  Therefore, the *best* parallelism CircleCI can achieve for us right now is bottlenecked on the slowest *file* of tests in the entire job.  The winner is...

```
JOB_TESTFILES: tests/pageslayout/test_journalist.py
tests/pageslayout/test_source.py
Autodetected filename timings.
RUN_TESTFILES: tests/pageslayout/test_journalist.py
███ Running SecureDrop application tests...
[...]
================= 68 passed, 14 warnings in 1130.34s (0:18:50) =================
```

That means that the best runtime we can achieve for `apptests` is currently `lint` (< 10 min) + `tests/pageslayout/test_journalist.py` (< 20 min + 5 min setup + 5 min validation) < 40 min.  Factoring out page-layout tests entirely with `parallelism: 2` achieved this runtime by explicitly putting this file in its own run, but so would the full `app-tests` job somewhere between [`parallelism: 5`](https://app.circleci.com/pipelines/github/freedomofpress/securedrop/2877/workflows/a166fe13-dfbe-4b57-991f-d26862dd0bb6/jobs/56293/parallel-runs/0?filterBy=ALL) and [`parallelism: 10`](https://app.circleci.com/pipelines/github/freedomofpress/securedrop/2883/workflows/57673efa-593c-4f92-876c-d9ff60d2435c/jobs/56285).


---

1. Actually, CircleCI can split on classname, too.  But:
    1. All of our Make → pytest logic is filename-based.
    2. Even if we were to add more support for class specifications, many of our `test_*.py` files contain only a single class anyway.

**Comment by cfm** [comment]
After #6077, latest `securedrop_ci` runs are down to ~43 minutes as expected with `app-page-layout-tests` at ~35 minutes.  Per https://github.com/freedomofpress/securedrop/issues/6026#issuecomment-905758983, no further optimization possible in current test layout.

**Comment by cfm** [comment]
After https://github.com/freedomofpress/securedrop/issues/6026#issuecomment-905758983, it's reasonable to conclude that `app-tests` job is as optimized as is currently possible given that the overall `securedrop_ci` workflow is now bottlenecked on #6077's factored-out `app-page-layout-tests`.  I've opened #6090 to track the optimization of the latter.
